### DISCUSSION TITLE
How to implement looping using the MIDI player ?

### DISCUSSION THREAD

**Discussion by bitdriftr** [FluidSynth/fluidsynth#777] [Q&A] [answered]
Hello,

I'm trying to loop a MIDI file from and to a specific tick using the Fluidsynth API.
I started from the [Loading and playing a MIDI file](https://www.fluidsynth.org/api/MIDIPlayer.html) code base, which works just as expected.

In order to call `fluid_player_seek()` and make the MIDI loop, I need a callback that is executed on every tick, which I didn't find.
I did find the `fluid_player_set_playback_callback()` function however :

* It is only called for MIDI events, so if there is no event on the tick I want to loop, the loop will be delayed until the next event.
* When it reaches the last tick (`fluid_player_get_current_tick() == fluid_player_get_total_ticks()`), the MIDI is removed from the play queue so audio is stopped despite a call to `fluid_player_seek(..., 0)`
* Even when adding the audio again using `fluid_player_add()`, I need to wait the next "call" to be able to use `fluid_player_seek()`, which is very inconvenient.

How can I implement this?

Thanks!

**Comment by derselbst** [comment]
The MIDI player hasn't been designed for such precise control. Feel free to propse changes to fix those quirks. However, I see no way of getting a callback on every tick.

Alternatively, you could consider to use the sequencer API instead. There, you can schedule timer events that give you a callback at a very certain tick. However you'll have to "emulate" the seeking yourself. I'm doing something like that together with libsmf in [my app](https://github.com/derselbst/ANMP/blob/master/src/InputLibraryWrapper/MidiWrapper.cpp), to  support MIDI track loops. But here I know when loop will happen as this is indicated by a CC. So not sure if this suits your use-case at all.

**Comment by bitdriftr** [comment, marked as answer]
Hi,

Thanks for your reply. I read the code in your app, and it indeed fixes the problem I was facing.
However, considering the amount of logic and additional state it creates, and considering most of this logic is already present in fluid player, I decided to implement a per-tick callback in the player.

I PR'd this in #780.
